File: app/core/backup.py
"""
追漫阁 - 数据备份与恢复
"""
import json
import logging
import os
import sqlite3
import tempfile
import hashlib
from datetime import datetime
import requests
from app import config
from app.db import database as db

logger = logging.getLogger(__name__)


def export_data() -> dict:
    """
    导出所有数据为字典

    Returns:
        包含所有追更数据的字典
    """
    animes = db.get_all_animes()
    result = {
        "version": "1.0",
        "exported_at": datetime.now().isoformat(),
        "app": "追漫阁",
        "animes": [],
        "settings": db.get_all_settings(),
    }

    for anime in animes:
        aid = anime["id"]
        episodes = db.get_episodes(aid)
        aliases = db.get_aliases(aid)
        rules = db.get_source_rules(aid)

        # 每集附带视频源
        ep_list = []
        for ep in episodes:
            sources = db.get_sources_for_episode(ep["id"])
            ep["sources"] = sources
            ep_list.append(ep)

        anime_data = {
            "anime": anime,
            "episodes": ep_list,
            "aliases": aliases,
            "rules": rules,
        }
        result["animes"].append(anime_data)

    return result


def export_json() -> str:
    """导出为 JSON 字符串"""
    data = export_data()
    return json.dumps(data, ensure_ascii=False, indent=2, default=str)


def import_data(data: dict) -> dict:
    """
    从字典导入数据（合并模式，不会删除现有数据）

    Args:
        data: export_data() 格式的字典

    Returns:
        导入结果统计
    """
    stats = {"animes_imported": 0, "episodes_imported": 0, "sources_imported": 0, "skipped": 0}

    for anime_entry in data.get("animes", []):
        anime_info = anime_entry.get("anime", {})
        tmdb_id = anime_info.get("tmdb_id")

        # 检查是否已存在
        existing = None
        if tmdb_id:
            existing = db.get_anime_by_tmdb_id(tmdb_id)

        if existing:
            anime_id = existing["id"]
            stats["skipped"] += 1
            # 更新观看进度（取较大值）
            if (anime_info.get("watched_ep") or 0) > (existing.get("watched_ep") or 0):
                db.update_anime(anime_id, {"watched_ep": anime_info["watched_ep"]})
        else:
            # 新增动漫
            anime_id = db.add_anime({
                "title_cn": anime_info.get("title_cn", ""),
                "title_en": anime_info.get("title_en", ""),
                "tmdb_id": tmdb_id,
                "poster_url": anime_info.get("poster_url", ""),
                "overview": anime_info.get("overview", ""),
                "air_date": anime_info.get("air_date", ""),
                "status": anime_info.get("status", ""),
                "total_episodes": anime_info.get("total_episodes", 0),
                "watched_ep": anime_info.get("watched_ep", 0),
            })
            stats["animes_imported"] += 1

        # 导入集数
        episodes_data = anime_entry.get("episodes", [])
        if episodes_data:
            existing_eps = {ep["absolute_num"] for ep in db.get_episodes(anime_id)}
            new_eps = []
            for ep in episodes_data:
                if ep.get("absolute_num") not in existing_eps:
                    new_eps.append({
                        "season_number": ep.get("season_number", 1),
                        "episode_number": ep.get("episode_number", 0),
                        "absolute_num": ep.get("absolute_num", 0),
                        "title": ep.get("title", ""),
                        "overview": ep.get("overview", ""),
                        "air_date": ep.get("air_date", ""),
                    })
            if new_eps:
                db.add_episodes(anime_id, new_eps)
                stats["episodes_imported"] += len(new_eps)

            # 恢复已看状态
            for ep in episodes_data:
                if ep.get("watched"):
                    db.mark_episode_watched(anime_id, ep["absolute_num"], True)

            # 导入视频源
            for ep in episodes_data:
                for src in ep.get("sources", []):
                    episode_obj = db.get_episode_by_num(anime_id, ep["absolute_num"])
                    if episode_obj:
                        try:
                            db.add_source({
                                "episode_id": episode_obj["id"],
                                "video_id": src.get("video_id", ""),
                                "title": src.get("title", ""),
                                "channel_name": src.get("channel_name", ""),
                                "channel_id": src.get("channel_id", ""),
                                "duration": src.get("duration", 0),
                                "view_count": src.get("view_count", 0),
                                "published_at": src.get("published_at", ""),
                                "match_score": src.get("match_score", 0),
                            })
                            stats["sources_imported"] += 1
                        except sqlite3.IntegrityError:
                            logger.debug(f"跳过重复视频源: {src.get('video_id', '')}")
                        except Exception as e:
                            logger.warning(f"导入视频源失败: {e}")

        # 导入别名
        for alias in anime_entry.get("aliases", []):
            try:
                db.add_alias(anime_id, alias)
            except sqlite3.IntegrityError:
                logger.debug(f"跳过重复别名: {alias}")
            except Exception as e:
                logger.warning(f"导入别名失败: {alias} - {e}")

        # 导入规则
        rules = anime_entry.get("rules")
        if rules:
            db.set_source_rules(anime_id, rules)

    # 导入设置
    for key, value in data.get("settings", {}).items():
        db.set_setting(key, str(value))

    return stats


def send_backup_to_telegram() -> dict:
    """
    将备份 JSON 通过 Telegram Bot 发送

    Returns:
        发送结果
    """
    # 优先从设置读取（UI 配置），回退到 config（环境变量）
    token = db.get_setting("tg_bot_token", "") or config.TG_BOT_TOKEN
    chat_id = db.get_setting("tg_chat_id", "") or config.TG_CHAT_ID

    if not token or not chat_id:
        error_msg = "未配置 TG_BOT_TOKEN 或 TG_CHAT_ID"
        logger.warning(error_msg)
        db.add_backup_log("telegram", "error", error_msg, 0, "", "NO_CONFIG")
        return {"success": False, "error": error_msg}

    tmp_path = None
    file_size = 0
    
    try:
        json_str = export_json()
        file_size = len(json_str.encode('utf-8'))
        now = datetime.now().strftime("%Y%m%d_%H%M%S")
        filename = f"zhuimange_backup_{now}.json"

        # 通过 sendDocument API 发送文件
        url = f"https://api.telegram.org/bot{token}/sendDocument"

        # 写入临时文件
        tmp = tempfile.NamedTemporaryFile(
            mode="w", suffix=".json", delete=False, encoding="utf-8"
        )
        tmp.write(json_str)
        tmp.close()
        tmp_path = tmp.name

        with open(tmp.name, "rb") as f:
            resp = requests.post(
                url,
                data={"chat_id": chat_id, "caption": f"📦 追漫阁备份 {now}"},
                files={"document": (filename, f, "application/json")},
                timeout=30,
            )

        if resp.status_code == 200 and resp.json().get("ok"):
            logger.info(f"Telegram 备份发送成功: {filename}")
            db.add_backup_log("telegram", "success", "备份发送成功", file_size, filename)
            return {"success": True, "filename": filename}
        else:
            error = resp.json().get("description", resp.text)
            error_code = resp.json().get("error_code", "TG_API_ERROR")
            logger.error(f"Telegram 备份发送失败: {error}")
            db.add_backup_log("telegram", "error", error, 0, filename, error_code)
            return {"success": False, "error": error}

    except requests.exceptions.Timeout:
        error_msg = "Telegram API 请求超时"
        logger.error(error_msg)
        db.add_backup_log("telegram", "error", error_msg, 0, "", "TIMEOUT")
        return {"success": False, "error": error_msg}
    except requests.exceptions.RequestException as e:
        error_msg = f"网络请求失败: {str(e)}"
        logger.error(f"Telegram 备份异常: {e}")
        db.add_backup_log("telegram", "error", error_msg, 0, "", "NETWORK_ERROR")
        return {"success": False, "error": error_msg}
    except Exception as e:
        error_msg = f"备份异常: {str(e)}"
        logger.error(f"Telegram 备份异常: {e}")
        db.add_backup_log("telegram", "error", error_msg, 0, "", "UNKNOWN_ERROR")
        return {"success": False, "error": str(e)}
    finally:
        if tmp_path and os.path.exists(tmp_path):
            try:
                os.unlink(tmp_path)
            except Exception as e:
                logger.warning(f"临时文件清理失败: {e}")


def calculate_backup_checksum(data: dict) -> str:
    """计算备份数据的校验和
    
    Args:
        data: 备份数据字典
    
    Returns:
        SHA256 哈希值
    """
    json_str = json.dumps(data, ensure_ascii=False, sort_keys=True, default=str)
    return hashlib.sha256(json_str.encode('utf-8')).hexdigest()


def verify_backup_integrity(data: dict, expected_checksum: str = None) -> dict:
    """验证备份数据的完整性
    
    Args:
        data: 备份数据字典
        expected_checksum: 期望的校验和（可选）
    
    Returns:
        验证结果字典
    """
    result = {
        "valid": True,
        "checksum": "",
        "checksum_match": False,
        "animes_count": 0,
        "episodes_count": 0,
        "sources_count": 0,
        "errors": []
    }
    
    # 检查基本结构
    if data.get("app") != "追漫阁":
        result["valid"] = False
        result["errors"].append("无效的备份文件标识")
    
    if "animes" not in data:
        result["valid"] = False
        result["errors"].append("缺少 animes 数据")
    else:
        result["animes_count"] = len(data["animes"])
    
    # 统计数据量
    episodes_count = 0
    sources_count = 0
    for anime_entry in data.get("animes", []):
        episodes = anime_entry.get("episodes", [])
        episodes_count += len(episodes)
        for ep in episodes:
            sources_count += len(ep.get("sources", []))
    
    result["episodes_count"] = episodes_count
    result["sources_count"] = sources_count
    
    # 计算校验和
    result["checksum"] = calculate_backup_checksum(data)
    
    # 验证校验和
    if expected_checksum:
        result["checksum_match"] = result["checksum"] == expected_checksum
        if not result["checksum_match"]:
            result["valid"] = False
            result["errors"].append("校验和不匹配")
    
    return result


def save_backup_local(backup_dir: str = None) -> dict:
    """保存备份到本地文件
    
    Args:
        backup_dir: 备份目录路径，默认为 data/backups
    
    Returns:
        备份结果
    """
    import os
    
    if not backup_dir:
        base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
        backup_dir = os.path.join(base_dir, "data", "backups")
    
    try:
        os.makedirs(backup_dir, exist_ok=True)
        
        data = export_data()
        json_str = json.dumps(data, ensure_ascii=False, indent=2, default=str)
        checksum = calculate_backup_checksum(data)
        
        now = datetime.now().strftime("%Y%m%d_%H%M%S")
        filename = f"zhuimange_backup_{now}.json"
        filepath = os.path.join(backup_dir, filename)
        
        with open(filepath, "w", encoding="utf-8") as f:
            f.write(json_str)
        
        file_size = os.path.getsize(filepath)
        
        logger.info(f"本地备份成功: {filename}, 大小: {file_size} 字节")
        db.add_backup_log("local", "success", "备份保存成功", file_size, filename)
        
        return {
            "success": True,
            "filename": filename,
            "filepath": filepath,
            "file_size": file_size,
            "checksum": checksum
        }
        
    except OSError as e:
        error_msg = f"文件系统错误: {str(e)}"
        logger.error(f"本地备份失败: {error_msg}")
        db.add_backup_log("local", "error", error_msg, 0, "", "FILESYSTEM_ERROR")
        return {"success": False, "error": error_msg}
    except Exception as e:
        error_msg = f"本地备份异常: {str(e)}"
        logger.error(f"本地备份异常: {e}")
        db.add_backup_log("local", "error", error_msg, 0, "", "UNKNOWN_ERROR")
        return {"success": False, "error": error_msg}

File: go/internal/backup/backup.go
package backup

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/csv"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"mime/multipart"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/lwhx/zhuimange/internal/model"
	"github.com/lwhx/zhuimange/internal/store"
)

// Service 封装备份导入导出依赖。
type Service struct {
	store *store.Store
	http  *http.Client
}

// ExportData 是 JSON 备份根结构。
type ExportData struct {
	Version    string            `json:"version"`
	ExportedAt string            `json:"exported_at"`
	App        string            `json:"app"`
	Settings   map[string]string `json:"settings"`
	Animes     []AnimeExport     `json:"animes"`
}

// AnimeExport 是单部动漫完整导出结构。
type AnimeExport struct {
	Anime    *model.Anime      `json:"anime"`
	Episodes []EpisodeExport   `json:"episodes"`
	Aliases  []string          `json:"aliases"`
	Rules    *model.SourceRule `json:"rules,omitempty"`
}

// EpisodeExport 是单集导出结构。
type EpisodeExport struct {
	Episode *model.Episode  `json:"episode"`
	Sources []*model.Source `json:"sources"`
}

// ImportStats 表示导入统计。
type ImportStats struct {
	AnimesImported   int `json:"animes_imported"`
	EpisodesImported int `json:"episodes_imported"`
	SourcesImported  int `json:"sources_imported"`
	Skipped          int `json:"skipped"`
}

var importableSettings = map[string]bool{
	"auto_sync_enabled":          true,
	"auto_sync_interval":         true,
	"match_threshold":            true,
	"match_recommend_threshold":  true,
	"invidious_url":              true,
	"invidious_fallback_urls":    true,
	"invidious_instance_weights": true,
	"tg_notify_enabled":          true,
	"tg_backup_enabled":          true,
	"tg_backup_interval_days":    true,
	"episode_sort_order":         true,
}

// NewService 创建备份服务。
func NewService(st *store.Store) *Service {
	return &Service{store: st, http: &http.Client{Timeout: 30 * time.Second}}
}

// ExportData 导出完整数据。
func (s *Service) ExportData(ctx context.Context) (*ExportData, error) {
	settings, err := s.store.GetAllSettings(ctx)
	if err != nil {
		return nil, err
	}
	animes, err := s.store.ListAnimes(ctx)
	if err != nil {
		return nil, err
	}
	result := &ExportData{Version: "1.0", ExportedAt: time.Now().Format(time.RFC3339), App: "追漫阁", Settings: settings, Animes: []AnimeExport{}}
	for _, anime := range animes {
		episodes, err := s.store.ListEpisodes(ctx, anime.ID)
		if err != nil {
			return nil, err
		}
		epExports := make([]EpisodeExport, 0, len(episodes))
		for _, episode := range episodes {
			sources, err := s.store.GetSourcesForEpisode(ctx, episode.ID)
			if err != nil {
				return nil, err
			}
			epExports = append(epExports, EpisodeExport{Episode: episode, Sources: sources})
		}
		aliases, err := s.store.GetAliases(ctx, anime.ID)
		if err != nil {
			return nil, err
		}
		rule, err := s.store.GetSourceRule(ctx, anime.ID)
		if err != nil {
			return nil, err
		}
		result.Animes = append(result.Animes, AnimeExport{Anime: anime, Episodes: epExports, Aliases: aliases, Rules: rule})
	}
	return result, nil
}

// ExportJSON 导出 JSON 字节。
func (s *Service) ExportJSON(ctx context.Context) ([]byte, error) {
	data, err := s.ExportData(ctx)
	if err != nil {
		return nil, err
	}
	return json.MarshalIndent(data, "", "  ")
}

// ImportJSON 从 JSON 合并导入。
func (s *Service) ImportJSON(ctx context.Context, payload []byte) (ImportStats, error) {
	var data ExportData
	if err := json.Unmarshal(payload, &data); err != nil {
		return ImportStats{}, err
	}
	stats := ImportStats{}
	for _, entry := range data.Animes {
		if entry.Anime == nil || strings.TrimSpace(entry.Anime.TitleCN) == "" {
			continue
		}
		anime := entry.Anime
		created, err := s.store.CreateAnime(ctx, anime)
		if err != nil {
			stats.Skipped++
			continue
		}
		stats.AnimesImported++
		for _, episodeEntry := range entry.Episodes {
			if episodeEntry.Episode == nil {
				continue
			}
			episode := *episodeEntry.Episode
			episode.AnimeID = created.ID
			inserted, err := s.store.AddEpisodes(ctx, []model.Episode{episode})
			if err == nil {
				stats.EpisodesImported += inserted
			}
			storedEpisode, err := s.store.GetEpisodeByNum(ctx, created.ID, episode.AbsoluteNum)
			if err != nil || storedEpisode == nil {
				continue
			}
			if episode.Watched {
				_ = s.store.SetEpisodeWatched(ctx, storedEpisode.ID, true)
			}
			for _, source := range episodeEntry.Sources {
				copySource := *source
				copySource.EpisodeID = storedEpisode.ID
				if err := s.store.AddSource(ctx, &copySource); err == nil {
					stats.SourcesImported++
				}
			}
		}
		for _, alias := range entry.Aliases {
			_ = s.store.AddAlias(ctx, created.ID, alias)
		}
		if entry.Rules != nil {
			entry.Rules.AnimeID = created.ID
			_ = s.store.UpsertSourceRule(ctx, entry.Rules)
		}
	}
	for key, value := range data.Settings {
		if !importableSettings[key] {
			continue
		}
		_ = s.store.SetSetting(ctx, key, value)
	}
	return stats, nil
}

// ExportCSV 导出通用追番列表 CSV。
func (s *Service) ExportCSV(ctx context.Context) ([]byte, error) {
	animes, err := s.store.ListAnimes(ctx)
	if err != nil {
		return nil, err
	}
	buffer := &bytes.Buffer{}
	writer := csv.NewWriter(buffer)
	_ = writer.Write([]string{"title_cn", "title_en", "tmdb_id", "bangumi_id", "watched_ep", "total_episodes", "status"})
	for _, anime := range animes {
		_ = writer.Write([]string{anime.TitleCN, anime.TitleEN, int64PtrToString(anime.TMDBID), int64PtrToString(anime.BangumiID), fmt.Sprintf("%d", anime.WatchedEp), fmt.Sprintf("%d", anime.TotalEpisodes), anime.Status})
	}
	writer.Flush()
	return buffer.Bytes(), writer.Error()
}

// ExportBangumi 导出 Bangumi 互通 JSON。
func (s *Service) ExportBangumi(ctx context.Context) ([]byte, error) {
	animes, err := s.store.ListAnimes(ctx)
	if err != nil {
		return nil, err
	}
	items := []map[string]any{}
	for _, anime := range animes {
		items = append(items, map[string]any{"bangumi_id": anime.BangumiID, "title": anime.TitleCN, "watched_ep": anime.WatchedEp, "total_episodes": anime.TotalEpisodes, "status": anime.Status})
	}
	return json.MarshalIndent(map[string]any{"version": "1.0", "items": items}, "", "  ")
}

// SendBackupToTelegram 发送 JSON 备份文件到 Telegram。
func (s *Service) SendBackupToTelegram(ctx context.Context) (map[string]any, error) {
	content, err := s.ExportJSON(ctx)
	if err != nil {
		return nil, err
	}
	token, _ := s.store.GetSetting(ctx, "tg_bot_token", "")
	chatID, _ := s.store.GetSetting(ctx, "tg_chat_id", "")
	if strings.TrimSpace(token) == "" || strings.TrimSpace(chatID) == "" {
		_ = s.store.AddBackupLog(ctx, "telegram", "error", "未配置 Telegram", 0, "", "NO_CONFIG")
		return nil, fmt.Errorf("未配置 Telegram")
	}
	filename := "zhuimange_backup_" + time.Now().Format("20060102_150405") + ".json"
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	_ = writer.WriteField("chat_id", chatID)
	_ = writer.WriteField("caption", "追漫阁备份 "+time.Now().Format("2006-01-02 15:04:05"))
	part, err := writer.CreateFormFile("document", filename)
	if err != nil {
		return nil, err
	}
	_, _ = part.Write(content)
	_ = writer.Close()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://api.telegram.org/bot"+token+"/sendDocument", body)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", writer.FormDataContentType())
	resp, err := s.http.Do(req)
	if err != nil {
		_ = s.store.AddBackupLog(ctx, "telegram", "error", err.Error(), 0, filename, "NETWORK_ERROR")
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		message := fmt.Sprintf("Telegram HTTP %d", resp.StatusCode)
		_ = s.store.AddBackupLog(ctx, "telegram", "error", message, 0, filename, "TG_API_ERROR")
		return nil, fmt.Errorf("%s", message)
	}
	checksum := sha256.Sum256(content)
	_ = s.store.AddBackupLog(ctx, "telegram", "success", "备份发送成功", int64(len(content)), filename, "")
	return map[string]any{"success": true, "filename": filename, "size": len(content), "sha256": hex.EncodeToString(checksum[:])}, nil
}

// SaveBackupLocal 将备份 JSON 保存到本地 data/backups 目录。
func (s *Service) SaveBackupLocal(ctx context.Context) (map[string]any, error) {
	backupDir := "data/backups"
	if err := os.MkdirAll(backupDir, 0o755); err != nil {
		return nil, fmt.Errorf("创建备份目录失败: %w", err)
	}

	content, err := s.ExportJSON(ctx)
	if err != nil {
		return nil, err
	}
	filename := "zhuimange_backup_" + time.Now().Format("20060102_150405") + ".json"
	fullPath := backupDir + string(os.PathSeparator) + filename
	if err := os.WriteFile(fullPath, content, 0o644); err != nil {
		_ = s.store.AddBackupLog(ctx, "local", "error", err.Error(), 0, filename, "FILESYSTEM_ERROR")
		return nil, err
	}
	checksum := sha256.Sum256(content)
	_ = s.store.AddBackupLog(ctx, "local", "success", "备份保存成功", int64(len(content)), filename, "")
	return map[string]any{
		"success":  true,
		"filename": filename,
		"filepath": fullPath,
		"size":     len(content),
		"sha256":   hex.EncodeToString(checksum[:]),
	}, nil
}

// int64PtrToString 将整数指针转字符串。
func int64PtrToString(value *int64) string {
	if value == nil {
		return ""
	}
	return fmt.Sprintf("%d", *value)
}
